extract_recipes.py

```diff
@@ -0,0 +1,220 @@
+
+        if len(qty) == 1:
+            qty = numeric(qty)
+        else:
+            try:
+                if '/' in qty:
+                    # 2 1/2
+                    qty = float(sum(Fraction(s) for s in qty.split()))
+                elif qty[-1].isdigit():
+                    # normal number, ending in [0-9]
+                    qty = float(qty)
+                else:
+                    # Assume the last character is a vulgar fraction
+                    qty = float(qty[:-1]) + numeric(qty[-1])
+            except ValueError:
+                pass # let it be a string
+    return qty
+
+# https://en.wikipedia.org/wiki/Cooking_weights_and_measures#United_States_measures
+measures = {
+    'drop': {'abrv': 'dr gt gtt', 'oz':1.0/576},
+    'smidgen': {'abrv': 'smdg smi', 'oz': 1.0/256},
+    'pinch': {'abrv': 'pn', 'oz': 1.0/128},
+    'dash': {'abrv': 'ds', 'oz': 1.0/64},
+    'saltspoon': {'abrv': 'ssp scruple', 'oz': 1.0/32},
+    'coffeespoon': {'abrv': 'csp', 'oz': 1.0/16},
+    'dram': {'abrv': 'dr', 'oz': 1.0/8},
+    'teaspoon': {'abrv': 'tsp t', 'oz': 1.0/6},
+    'tablespoon': {'abrv': 'Tbsp T', 'oz': 1.0/2},
+    'ounce': {'abrv': 'oz fl.oz', 'oz': 1.0},
+    'wineglass': {'abrv': 'wgf', 'oz': 2.0},
+    'teacup': {'abrv': 'tcf gill', 'oz': 4.0},
+    'cup': {'abrv': 'C', 'oz': 8.0},
+    'pint': {'abrv': 'pt', 'oz': 16.0},
+    'quart': {'abrv': 'qt', 'oz': 32.0},
+    'pottle': {'abrv': 'pot', 'oz': 64.0},
+    'gallon': {'abrv': 'gal', 'oz': 128.0},
+    'pound': {'abrv': 'lbs', 'oz': 16.0},
+    'gram': {'abrv': 'g', 'oz': 0.035274},
+    'kilogram': {'abrv': 'kg', 'oz': 35.274},
+}
+
+def normalize_unit(unit):
+    if unit is not None:
+        for u, d in measures.items():
+            if unit.lower() == u:
+                return u
+            if unit in d['abrv'].split(' '):
+                return u
+    return unit
+
+def normalize_name(name):
+    return capwords(name)        .strip(' \t\n\r,.')        .replace('Whole ', '')        .replace('Half ', '')        .replace('Hot ', '')        .replace('Warm ', '')        .replace('Cold ', '').strip()
+
+def parse_ingredient(i):
+    if type(i) is not str:
+        if i.string is None:
+            # multiple tags in child,
+            # bs4 gets confused per https://www.crummy.com/software/BeautifulSoup/bs4/doc/#string
+            s = ' '.join(i.stripped_strings)
+        else:
+            s = i.string.strip()
+    else:
+        s = i
+    raw = s.replace('\xa0','').strip()
+    
+    clean = re.sub(r'\(.+?\)', '', raw).replace('’',"'")
+
+    parsed = pattern.match(clean)
+    if parsed:
+        qty, unit, name = parsed.groups()
+    else:
+        print("WARN: Unable to parse ingredient '{}'".format(raw))
+        return Ingredient(None, None, None, raw)
+    
+    qty = normalize_qty(qty)
+            
+    unit = normalize_unit(unit)
+
+    name = normalize_name(name)
+
+    return Ingredient(qty, unit, name, raw)
+
+# Tests to validate parse_ingredient()!!!
```

Comment:
Could you please add a TODO related to your comment about these tests needing to be pulled out into their own file?